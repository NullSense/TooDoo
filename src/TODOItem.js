import React from 'react';
import PropTypes from 'prop-types';

/**
 * This component represents a TODOItem
 * @version 0.1
 * @author salsa20
 * @extends React.Component
 * @param {} props pass on parent data
 */
class TODOItem extends React.Component {
  constructor(props) {
    super(props);

    // local item state
    this.state = {
      entry: '',
      modifiable: true
    };

    // bind context
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyPressed = this.handleKeyPressed.bind(this);
  }

  /**
   * on input update local entry
   * @param {} event pass on input event
   */
  handleChange(event) {
    this.setState({
      entry: event.target.value
    });
  }

  /**
   * call top level delete function passed on as prop
   * @param {} event pass on input event
   */
  handleDelete(event) {
    event.preventDefault();
    this.props.del(this.props.id);
  }

  /**
   * on enter disallow input
   * @param {} event pass on input event
   */
  handleKeyPressed(event) {
    // if enter was pressed
    if (event.keyCode === 13) {
      // render differently, send to server
      this.setState({
        modifiable: false
      });
    } else if (event.button === 0) {
      this.setState({
        modifiable: true
      });
    }
  }

  render() {
    // if modfiable render input field, else display list item with delete button
    if (!this.state.modifiable) {
      return (
        <form onSubmit={this.handleDelete.bind(this)}>
          <div className="EntryWrapper">
            <li className="TODOItem" onClick={this.handleKeyPressed}>
              {this.state.entry}
            </li>
            <button className="DelEntry" type="submit">
              <img
                style={{ filter: 'opacity(50%)', maxWidth: '50%', height: 'auto' }}
                src="https://img.icons8.com/metro/26/000000/trash.png"
                alt="-"
              />
            </button>
          </div>
        </form>
      );
    } else {
      return (
        <form className="EntryForm">
          <li>
            <textarea
              className="TODOItem"
              value={this.state.entry}
              onChange={this.handleChange}
              onKeyDown={this.handleKeyPressed}
              placeholder="type a note..."
            />
          </li>
        </form>
      );
    }
  }
}

// define prop types for TODOItem
TODOItem.propTypes = {
  del: PropTypes.func,
  id: PropTypes.number
};

export default TODOItem;
