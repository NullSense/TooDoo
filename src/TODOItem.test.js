import { configure, mount, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import TODOItem from './TODOItem.js';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });

// call this to mount the component by passing on props
const setUp = props => {
  const component = mount(<TODOItem {...props} />);
  return component;
};

describe('TODOItem', () => {
  // always setup a component with empty props
  let component;
  beforeEach(() => {
    component = mount(<TODOItem />);
  });

  // unmount after each test
  afterEach(() => {
    component.unmount();
  });

  /*
   * snapshot testing
   */
  it('renders correctly', () => {
    const snapshot = renderer.create(<TODOItem />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const snapshot = renderer.create(<TODOItem key={Date.now()} id={Date.now()} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  // Simulate keyboard input
  it('changes state correctly after hitting enter', () => {
    // simulate hitting enter on a new item
    component.find('textarea').simulate('keydown', { keyCode: 13 });
    expect(component.state().modifiable).toBeFalsy();
    expect(component.state().modifiable).toEqual(false);

    // simulate then clicking on the item again
    component.find('.TODOItem').simulate('click', { button: 0 });
    expect(component.state().modifiable).toBeTruthy();
    expect(component.state().modifiable).toEqual(true);
  });

  // pass on an id and match
  it('should have correct id', () => {
    component = setUp({ id: 1, entry: 'abc' });
    expect(component.props().id).toEqual(1);
  });

  // test for initial modifiability
  it('should be modifiable on mount', () => {
    component = setUp({ id: 1, entry: 'abc' });
    expect(component.state().modifiable).toBeTruthy();
    expect(component.state().modifiable).toBe(true);
  });

  // on mount the component shouldn't have an entry
  it('should have an empty entry on mount', () => {
    expect(component.state().entry).toEqual('');
  });

  // check if delete function would get called
  it('should run the passed on delete function', () => {
    const mockDel = jest.fn();
    component = setUp({ del: mockDel });
    component.instance().handleDelete({ preventDefault() {} });
    expect(mockDel).toBeCalled();
  });
});
