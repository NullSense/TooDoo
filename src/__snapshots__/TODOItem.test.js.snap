// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TODOItem renders correctly 1`] = `
<form
  className="EntryForm"
>
  <li>
    <textarea
      className="TODOItem"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="type a note..."
      value=""
    />
  </li>
</form>
`;

exports[`TODOItem renders correctly 2`] = `
<form
  className="EntryForm"
>
  <li>
    <textarea
      className="TODOItem"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="type a note..."
      value=""
    />
  </li>
</form>
`;
